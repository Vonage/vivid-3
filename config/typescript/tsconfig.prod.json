{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "../../dist",
    "baseUrl": "../../src"
  },
  "exclude": [
    "../../src/**/*.spec.ts",
    "../../src/**/*.test.ts"
  ]
}
