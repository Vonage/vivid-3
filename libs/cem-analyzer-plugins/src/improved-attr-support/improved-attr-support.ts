import type {
	AnalyzePhaseParams,
	Plugin,
} from '@custom-elements-manifest/analyzer';

/**
 * This plugin fixes an issue with the @attr decorator plugin.
 * When using @attr with options but no explicit attribute name, e.g. `@attr({ mode: 'boolean' })`, the plugin will not
 * default the attribute name to the fieldName, and instead it will be missing.
 * This leads to issues downstream when processing inheritance, which expects attributes to have a name.
 */
export const improvedAttrSupportPlugin = (): Plugin => ({
	name: 'improved-attr-support',
	analyzePhase({ ts, node, moduleDoc }: AnalyzePhaseParams) {
		switch (node.kind) {
			case ts.SyntaxKind.ClassDeclaration: {
				const className = (node as any).name?.getText();
				const classDoc = moduleDoc?.declarations?.find(
					(declaration) => declaration.name === className
				);
				for (const attr of (classDoc as any)?.attributes ?? []) {
					if (!attr.name) {
						attr.name = attr.fieldName;
					}
				}
				break;
			}
		}
	},
});
