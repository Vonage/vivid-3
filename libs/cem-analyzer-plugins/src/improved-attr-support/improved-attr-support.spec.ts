import * as path from 'path';
import * as fs from 'fs';
import * as glob from 'glob';
import * as ts from '@custom-elements-manifest/analyzer/node_modules/typescript';
import { create } from '@custom-elements-manifest/analyzer/src/create.js';
import { fastPlugin } from '@custom-elements-manifest/analyzer/src/features/framework-plugins/fast/fast.js';
import { improvedAttrSupportPlugin } from './improved-attr-support';

const fixturesPath = path.join(__dirname, '__fixtures__');
const testCases = fs.readdirSync(fixturesPath);

describe('improvedAttrSupportPlugin', () => {
	it.each(testCases)(
		`should produce correct results for testcase %s`,
		async (testCase) => {
			const moduleFiles = glob.sync(
				path.join(fixturesPath, testCase, '/**/*.{ts,js}')
			);
			const modules = moduleFiles.map((moduleFile) =>
				ts.createSourceFile(
					'my-element.js',
					fs.readFileSync(moduleFile, 'utf-8'),
					ts.ScriptTarget.ES2015,
					true
				)
			);

			expect(
				create({
					modules,
					plugins: [...fastPlugin(), improvedAttrSupportPlugin()],
				})
			).toMatchSnapshot();
		}
	);
});
