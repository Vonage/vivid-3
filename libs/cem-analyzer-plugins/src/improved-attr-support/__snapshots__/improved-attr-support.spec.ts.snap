// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`improvedAttrSupportPlugin > should produce correct results for testcase default 1`] = `
{
  "modules": [
    {
      "declarations": [
        {
          "attributes": [
            {
              "default": "''",
              "fieldName": "noOptions",
              "name": "noOptions",
              "type": {
                "text": "string",
              },
            },
            {
              "default": "false",
              "fieldName": "optionsWithoutName",
              "name": "optionsWithoutName",
              "type": {
                "text": "boolean",
              },
            },
            {
              "default": "false",
              "fieldName": "optionsWithName",
              "name": "specified name",
              "type": {
                "text": "boolean",
              },
            },
          ],
          "customElement": true,
          "description": "",
          "kind": "class",
          "members": [
            {
              "default": "''",
              "kind": "field",
              "name": "noOptions",
              "type": {
                "text": "string",
              },
            },
            {
              "default": "false",
              "kind": "field",
              "name": "optionsWithoutName",
              "type": {
                "text": "boolean",
              },
            },
            {
              "default": "false",
              "kind": "field",
              "name": "optionsWithName",
              "type": {
                "text": "boolean",
              },
            },
          ],
          "name": "MyComponent",
          "superclass": {
            "name": "HTMLElement",
          },
        },
      ],
      "exports": [
        {
          "declaration": {
            "module": "my-element.js",
            "name": "MyComponent",
          },
          "kind": "js",
          "name": "MyComponent",
        },
      ],
      "kind": "javascript-module",
      "path": "my-element.js",
    },
  ],
  "readme": "",
  "schemaVersion": "1.0.0",
}
`;
